[6X]Utility tool for EL7 -> EL8 OS upgrade

z-wenlin opened the pull request:
After upgrading from RHEL7 to RHEL8, the locale data changes due to Glibc upgrade.
Please refer <https://wiki.postgresql.org/wiki/Locale_data_changes>.

This PR adds a utility tool for prechecking and fixing indexes and tables/partitions.

It includes `precheck-index`, `precheck-table`, and `migrate` commands.
1. `precheck-index` is used for detecting indexes under built-in collatable data types（Like text, varchar, and char) that need to be reindexed (including catalog and user indexes).
2. `precheck-table` is used for detecting range-partitioned tables using those types in the partition key, which should be checked to verify that all rows are still in the correct partitions after the OS upgrade. Also checked the tables with the distribution key using the custom operator class. 
    Before the OS upgrade, `precheck-table` should with the option `--pre_upgrade`, it will list the partition tables with the partition key using built-in collatable types.  After the OS upgrade, should run `precheck-table` again without the option, it will do another GUC check to verify the rows are still in the correct partitions after the OS upgrade.
3. `migrate` is used to handle the above indexes and tables. If needs to reindex, then reindex. If needs repartition, then re-partition the tables. 

------

For the implementation of `precheck-index`, we use the SQL(Provided by Postgres) to get which indexes are affected.

For the implementation of `precheck-table`, we firstly use SQL to filter the partition tables using built-in collatable types in the partition key, then using the GUC of `gp_detect_data_correctness`, which was introduced by #16367, execute the query of `insert leaf_partition_table select * from leaf_partition_table` for all the leaf partition tables, to verify if the rows are in the correct partition, if the check failed, output them to the specified file. During this process, we use multiple threads to run the detection concurrently, since after we set the GUC of `gp_detect_data_correctness`, data will not insert into the leaf table, we don't need to consider the disk space, we can use 50 or more threads to run it.

For the implementation of `migrate`, we iterate all the affected indexes and tables. For the index, just use the statement `reindex index xxx` to rebuild. For the table, we use the following transaction to replace the original table:

```
begin; 
create temp table xxx_bak as select * from xxx; 
truncate xxx; 
insert into xxx select * from xxx_bak; 
commit;
```

SmartKeyerror commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,351 @@
+WHERE collname NOT IN ('C', 'POSIX') and indexrelid >= 16384;
+        """
+        index = db.query(sql).getresult()
+        logger.info("There are %d user indexes in database %s that might be affected due to upgrade.", len(index), dbname)
```
I think it's time to use `.format()` to replace the old `%` comprehensively, cause the "%" format for Python has a very long history. Besides, I believe `.format()` is more convenient and portable for higher Python versions.

In Python 3.6, we can even use f-format, it's super convenient and readable:

```
 >>> a = "biubiubiu"
>>> b = f"xiuxiuxiu, {a}"
>>> b
'xiuxiuxiu, biubiubiu'
```

z-wenlin replied:
Thanks for the reminder, have replaced it by using .format()

SmartKeyerror commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,351 @@
+        self.parse_inputfile()
+
+        import json
+        logger.info(json.dumps(self.dbdict))
```
This could be deleted

z-wenlin replied:
Done.

z-wenlin commented:
> Does 7X need this function too? If so, we should merge it into the main branch, and then backport to 6X.

Thanks for the review. I think this utility tool is only used for the 6X and no need for 7X so far.

charliettxx commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,400 @@
+        FROM 
+        might_affected_tables group by (prelid, coll, parisdefault)
+        )
+        select prelid::regclass::text as partitionname, coll, bool_or(parisdefault) as parhasdefault from par_has_default group by (prelid, coll) ;
```
why do we `cast prelid::regclass::text `here? Is there any function transfer prelid to partitionname? I guess you want to get partitionname by prelid for this column here

z-wenlin replied:
The reason for `cast prelid::regclass::text ` is that I need to output the `partitiontablename`.
There are other ways to get the name, for example, in the `pg_partitions` it has the `partitionschemaname` and `partitiontablename`, but this is a view, and it joins pg_class to get `relname`. I try to avoid using it, so I used the `parrelid` and cast it, and it also contains the schema name.

charliettxx commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,400 @@
+        SELECT indexrelid::regclass::text, indrelid::regclass::text, coll, collname, pg_get_indexdef(indexrelid)
+FROM (SELECT indexrelid, indrelid, indcollation[i] coll FROM pg_index, generate_subscripts(indcollation, 1) g(i)) s
+JOIN pg_collation c ON coll=c.oid
+WHERE collname NOT IN ('C', 'POSIX') and indexrelid >= 16384;
```
I suppose for `NOT IN ('C', 'POSIX'),` it's better to use `collname != 'C' and collname != 'POSIX' ` instead of NOT IN expression.

z-wenlin replied:
Done.

charliettxx commented:
looks good to me on the test cases and SQL part, not digging into python part.

SmartKeyerror commented:
I noticed this PR doesn't use the GUC of `gp_detect_data_correctness`, which I added in #16367, so do we get the problem tables?

z-wenlin commented:
> I noticed this PR doesn't use the GUC of `gp_detect_data_correctness`, which I added in #16367, so do we get the problem tables?

Yes, this PR was raised before #16367 was merged, and before that, we decided to list all the partition tables that might be affected due to the upgrade, and it used the following method to check the problem tables.

```
begin; create temp table XXX_bak as select * from XXX; truncate XXX; insert into XXX select * from XXX_bak; commit;
```

SmartKeyerror commented:
> > I noticed this PR doesn't use the GUC of `gp_detect_data_correctness`, which I added in #16367, so do we get the problem tables?
> 
> Yes, this PR was raised before #16367 was merged, and before that, we decided to list all the partition tables that might be affected due to the upgrade, and it used the following method to check the problem tables.
> 
> ```
> begin; create temp table XXX_bak as select * from XXX; truncate XXX; insert into XXX select * from XXX_bak; commit;
> ```

Does that mean we need to re-write all the partition tables which is potentially problematic?

z-wenlin commented:
> > > I noticed this PR doesn't use the GUC of `gp_detect_data_correctness`, which I added in #16367, so do we get the problem tables?
> > 
> > 
> > Yes, this PR was raised before #16367 was merged, and before that, we decided to list all the partition tables that might be affected due to the upgrade, and it used the following method to check the problem tables.
> > ```
> > begin; create temp table XXX_bak as select * from XXX; truncate XXX; insert into XXX select * from XXX_bak; commit;
> > ```
> 
> Does that mean we need to re-write all the partition tables which is potentially problematic?

Currently, yes, for all partition range tables, we will try to insert and see if there are errors. 
Besides, I have updated the PR to use the GUC `gp_detect_data_correctness`, please help to review,  thanks.

SmartKeyerror commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,380 @@
+        start = time.time()
+        db.query("set gp_detect_data_correctness = on;")
+        logger.info("db: {}, executing command: {}".format(db_name, command))
+        db.query(command)
```
If we don't process the query result, how do we know whether the current partition table should be rebuilt or not?

Because if we set the GUC `gp_detect_data_correctness` to `on`, an ERROR will be thrown when we execute `insert xx select * from xx` if the partition or distribution is not correct, then output those tables in another file, and let the customer process them, am I right?

SmartKeyerror replied:
So I think `gp_detect_data_correctness` should be used at `CheckTables.dump()`.

Firstly, we use `self.get_affected_partitioned_tables(dbname)` to fetch all possible affected partition tables, and use the GUC `gp_detect_data_correctness` to verify it, if an error is thrown, just add the table into the output files.

After all, execute `insert into ... select * from ...` to reorganize data really.

z-wenlin replied:
Yes, it makes sense. Currently, the precheck is to list all the might-affected range-partition tables and then try to ` insert XXX select * from XXX` to see if the partition is correct and outputs all the results into the report.
I will modify the PR and use the GUC to filter the tables again.

SmartKeyerror commented on gpMgmt/bin/gpupgradecheck/README.md:
```diff
@@ -0,0 +1,164 @@
+1. use `python upgrade_check.py precheck-index` to list affected indexes.
+2. use `pythonn upgrade_check.py precheck-table` to list affected partitioned tables.
```
typo, "pythonn" -> "python"

SmartKeyerror commented on gpMgmt/bin/gpupgradecheck/README.md:
```diff
@@ -0,0 +1,164 @@
+optional arguments:
+  -h, --help     show this help message and exit
+  --input INPUT  the file contains reindex or rebuild partition ccommandsmds
+  --nproc NPROC  the concurrent proces to run the commands
```
The optional argument of `--nproc` should be put on the `python upgrade_check.py precheck-table -h`, and please change another name for it.

SmartKeyerror commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,465 @@
+
+dbkeywords = "-- DB name: "
+
+def sig_handler(sig, arg):
```
Do we still need this function?

bmdoil reviewed (commented):
> For the implementation of precheck-index, we just use a simple SQL to filter the probably be affected indexes, not 100% accurate.

What is preventing our check from being 100% accurate? How helpful is the tool for users if the detection method isn't 100% accurate?

z-wenlin commented:
> > For the implementation of precheck-index, we just use a simple SQL to filter the probably be affected indexes, not 100% accurate.
> 
> What is preventing our check from being 100% accurate? How helpful is the tool for users if the detection method isn't 100% accurate?

The check is accurate. For Index, the SQL we use is consistent with PG’s official documentation. The previous description is inaccurate and the Description has been updated.

Thanks for the review, the comments have been fixed.

z-wenlin commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,483 @@
+        """
+        index = db.query(sql).getresult()
+        if index:
+            logger.info("There are {} user indexes in database {} that might be affected due to upgrade.".format(len(index), dbname))
```
Yes, agreed, I changed it to `gp_upgrade_el8_locale_check`

kalensk replied:
I think using the term `gp_upgrade_` is still very confusing both for customers and internally since as @bmdoil mentioned we don't want to confuse this with `gpupgrade` which does in-place GPDB upgrades and this tool is for OS upgrades. 

That is, we want to be very clear that this tool is for operating system upgrades, and gpupgrade is for GPDB upgrades.

Thus, I think prefixing all terms in this code base with "operating system" or OS upgrade is the way to go for clarity. So something like `os_upgrade_el8_locale_check`. Similarly throughout all the documentation and README's. For example, in the README I noticed:
> 2023-10-18 11:04:13,944 - INFO - There are 2 catalog indexes that needs reindex when doing in-place upgrade from EL7->EL8.

To me I find this line subtlety misleading/confusing and suggest adding the adjective "operating system" to the term upgrade:
> 2023-10-18 11:04:13,944 - INFO - Found 2 catalog indexes that need to be reindexed when doing operating system upgrade from EL7->EL8.

z-wenlin replied:
Thanks for the suggestion, I rename it to `el8_migrate_locale`

z-wenlin commented on gpMgmt/bin/gpupgradecheck/upgrade_check.py:
```diff
@@ -0,0 +1,483 @@
+        # The built-in collatable data types are text,varchar,and char, and the indcollation contains the OID of the collation 
+        # to use for the index, or zero if the column is not of a collatable data type.
+        sql = """
+        SELECT indexrelid::regclass::text, indrelid::regclass::text, coll, collname, pg_get_indexdef(indexrelid)
```
Nice catch, I modified the SQL and add `distinct(indexrelid)` to filter the duplicate entries.

kalensk commented on gpMgmt/bin/gp_upgrade_el8_locale_check/README.md:
```diff
@@ -0,0 +1,213 @@
+1. use `python upgrade_el8_locale_check.py precheck-index` to list affected indexes.
+2. use `python upgrade_el8_locale_check.py precheck-table` to list affected partitioned tables.
+3. use `python upgrade_el8_locale_check.py postfix` to run the reindex and alter partition table commands.
```
Side note in terms of the API I would have gone more with something like `list table` and `list index` where the list command can take an argument of table/index/etc. 

I would have considered renaming the `postfix` command to something like `migrate` since its both reindexing and altering tables. 

This means the name of the tool isn't quite accurate since its not just checking, but also migrating, reindexing, and altering. A name like `os_upgrade_migrate_locale` or similar is more descriptive and accurate.

bmdoil commented on gpMgmt/bin/el8_migrate_locale/README.md:
```diff
@@ -0,0 +1,213 @@
+
+2023-10-16 04:12:19,804 - WARNING - no default partition for testupgrade.partition_range_test_3
+...
+2023-10-16 04:12:22,058 - INFO - Current progress: have 0 remaining, 2.99126505852 seconds passed.
+2023-10-16 04:12:22,058 - INFO - worker[0]: finish.
+---------------------------------------------
+total table size (in GBytes) : 0.000396907329559
```
nit: `2.99126505852 seconds passed` this amount of precision seems unnecessary. Can we just format it to x.xx seconds?

nit: `total table size (in GBytes) : 0.000396907329559` for values this small we should present as `396 kB` or omit altogether.

bmdoil commented on gpMgmt/bin/el8_migrate_locale/el8_migrate_locale.py:
```diff
@@ -0,0 +1,502 @@
+                port = self.get_port_from_conf()
+            return int(port)
+        except:
+            sys.exit("No port been set, please set env PGPORT or MASTER_DATA_DIRECTORY or specify the port in the command line")
```
nit: `No port has been set`

bmdoil commented on gpMgmt/bin/el8_migrate_locale/el8_migrate_locale.py:
```diff
@@ -0,0 +1,502 @@
+        """
+        index = db.query(sql).getresult()
+        if index:
+            logger.info("There are {} catalog indexes that needs reindex when doing OS upgrade from EL7->EL8.".format(len(index)))
```
For 6X is there ever a case where the catalog indexes differ from this:
```
pg_shseclabel_object_index
pg_shseclabel_object_index
```
They're all declared in `src/include/catalog/indexing.h` and users aren't going to be modifying system table indexes right?

z-wenlin replied:
I think users usually won't modify the system table index.  On 6X, for catalog indexes, it's always these two indexes, the column provider is using ` text_ops` and its collation is not C or POSIX, which is why it's filtered out. 
```
DECLARE_UNIQUE_INDEX(pg_seclabel_object_index, 3597, on pg_seclabel using btree(objoid oid_ops, classoid oid_ops, objsubid int4_ops, provider text_ops));
#define SecLabelObjectIndexId				3597
DECLARE_UNIQUE_INDEX(pg_shseclabel_object_index, 3593, on pg_shseclabel using btree(objoid oid_ops, classoid oid_ops, provider text_ops));
#define SharedSecLabelObjectIndexId			3593
```
```

postgres=# select * from pg_index where indexrelid= 3597;
 indexrelid | indrelid | indnatts | indisunique | indisprimary | indisexclusion | indimmediate | indisclustered | indisvalid | indcheckxmin | indisready | indislive | indi
sreplident | indkey  | indcollation |      indclass       | indoption | indexprs | indpred
------------+----------+----------+-------------+--------------+----------------+--------------+----------------+------------+--------------+------------+-----------+-----
-----------+---------+--------------+---------------------+-----------+----------+---------
       3597 |     3596 |        4 | t           | f            | f              | t            | f              | t          | f            | t          | t         | f
           | 1 2 3 4 | 0 0 0 100    | 1981 1981 1978 3126 | 0 0 0 0   |          |
(1 row)
postgres=# select oid, * from pg_collation ;
 oid | collname | collnamespace | collowner | collencoding | collcollate | collctype
-----+----------+---------------+-----------+--------------+-------------+-----------
 100 | default  |            11 |        10 |           -1 |             |
 950 | C        |            11 |        10 |           -1 | C           | C
 951 | POSIX    |            11 |        10 |           -1 | POSIX       | POSIX
(3 rows)
```
But on 7X, I find the collation for these tables has been changed to C.
```
postgres=# \dS+ pg_shseclabel
                             Table "pg_catalog.pg_shseclabel"
  Column  | Type | Collation | Nullable | Default | Storage  | Stats target | Description
----------+------+-----------+----------+---------+----------+--------------+-------------
 objoid   | oid  |           | not null |         | plain    |              |
 classoid | oid  |           | not null |         | plain    |              |
 provider | text | C         | not null |         | extended |              |
 label    | text | C         | not null |         | extended |              |
Indexes:
    "pg_shseclabel_object_index" UNIQUE, btree (objoid, classoid, provider), tablespace "pg_global"
Tablespace: "pg_global"
Access method: heap
postgres=# select indcollation from pg_index where indexrelid=3593;
 indcollation
--------------
 0 0 950
(1 row)
```
So I think it's better to print it out.

bmdoil commented on gpMgmt/bin/el8_migrate_locale/el8_migrate_locale.py:
```diff
@@ -0,0 +1,502 @@
+        # The built-in collatable data types are text,varchar,and char, and the indcollation contains the OID of the collation 
+        # to use for the index, or zero if the column is not of a collatable data type.
+        sql = """
+        SELECT distinct(indexrelid), indexrelid::regclass::text as indexname, indrelid::regclass::text as tablename, collname, pg_get_indexdef(indexrelid)
```
Depending on the value of `search_path` the `indexname` and `tablename` might not include the schema. Since were using the 'indexname' to issue the `reindex index` command.

I think we should make sure it is fully qualified even if it makes the query more expensive. If the default `search_path` happens to change between when `precheck-index` is run and `migrate` is run, the `migrate` will fail. 

I ran `test.sql` into schemas `public` `foo` and `bar` with default `search_path`, then set `search_path` default to `bar` with `gpconfig` and a restart.

```
$ psql -c "show search_path" 
 search_path 
-------------
 'bar'
```
```
python el8_migrate_locale.py migrate --input index.out 

2023-10-19 15:37:42,345 - INFO - db: postgres, executing command: reindex index foo." test_id "" 3 ";
2023-10-19 15:37:42,359 - INFO - db: postgres, executing command: reindex index foo.idx_projecttag;
2023-10-19 15:37:42,371 - INFO - db: postgres, executing command: reindex index "test_id2 \ $ \\";
2023-10-19 15:37:42,372 - ERROR - ERROR:  relation "test_id2 \ $ \\" does not exist

2023-10-19 15:37:42,376 - INFO - db: postgres, executing command: reindex index foo.id1;
2023-10-19 15:37:42,389 - INFO - db: postgres, executing command: reindex index test_idx_citext;
2023-10-19 15:37:42,389 - ERROR - ERROR:  relation "test_idx_citext" does not exist

2023-10-19 15:37:42,393 - INFO - db: postgres, executing command: reindex index bar."test_id1 's ";
2023-10-19 15:37:42,406 - INFO - db: postgres, executing command: reindex index bar."test_id2 \ $ \\";
2023-10-19 15:37:42,418 - INFO - db: postgres, executing command: reindex index idx_projecttag;
2023-10-19 15:37:42,419 - ERROR - ERROR:  relation "idx_projecttag" does not exist

2023-10-19 15:37:42,423 - INFO - db: postgres, executing command: reindex index " test_id "" 3 ";
2023-10-19 15:37:42,423 - ERROR - ERROR:  relation " test_id " 3 " does not exist

2023-10-19 15:37:42,427 - INFO - db: postgres, executing command: reindex index bar.id1;
2023-10-19 15:37:42,439 - INFO - db: postgres, executing command: reindex index foo."test_id2 \ $ \\";
2023-10-19 15:37:42,452 - INFO - db: postgres, executing command: reindex index bar." test_id "" 3 ";
2023-10-19 15:37:42,464 - INFO - db: postgres, executing command: reindex index foo."test_id1 's ";
2023-10-19 15:37:42,477 - INFO - db: postgres, executing command: reindex index test_citext_pkey;
2023-10-19 15:37:42,477 - ERROR - ERROR:  relation "test_citext_pkey" does not exist

2023-10-19 15:37:42,481 - INFO - db: postgres, executing command: reindex index multi_col_idx;
2023-10-19 15:37:42,482 - ERROR - ERROR:  relation "multi_col_idx" does not exist

2023-10-19 15:37:42,485 - INFO - db: postgres, executing command: reindex index "test_id1 's ";
2023-10-19 15:37:42,486 - ERROR - ERROR:  relation "test_id1 's " does not exist

2023-10-19 15:37:42,490 - INFO - db: postgres, executing command: reindex index bar.idx_projecttag;
2023-10-19 15:37:42,499 - INFO - All done
```

z-wenlin replied:
> If the default search_path happens to change between when precheck-index is run and migrate is run, the migrate will fail.

For this scenario, I think we should suggest users don't change the search_path between precheck and migrate.
If that does happen, users could run precheck again, then migrate, and no fail would happen.
This is the test result.

```
$ psql -c "show search_path" -d postgres
 search_path
-------------
 bar
(1 row)
$ python el8_migrate_locale.py precheck-index --out index.out
2023-10-20 17:03:28,866 - INFO - There are 2 catalog indexes that needs reindex when doing OS upgrade from EL7->EL8.
2023-10-20 17:03:28,902 - INFO - There are 1 user indexes in database postgres that needs reindex when doing OS upgrade from EL7->EL8.
2023-10-20 17:03:28,923 - INFO - There are 7 user indexes in database test1 that needs reindex when doing OS upgrade from EL7->EL8.
2023-10-20 17:03:28,945 - INFO - There are 7 user indexes in database test2 that needs reindex when doing OS upgrade from EL7->EL8.
2023-10-20 17:03:28,967 - INFO - There are 7 user indexes in database test3 that needs reindex when doing OS upgrade from EL7->EL8.

$ python el8_migrate_locale.py migrate --input index.out
2023-10-20 17:03:47,463 - INFO - db: test1, total have 7 commands to execute
2023-10-20 17:03:47,468 - INFO - db: test1, executing command: reindex index public.test_citext_pkey;
2023-10-20 17:03:47,484 - INFO - db: test1, executing command: reindex index public.test_idx_citext;
2023-10-20 17:03:47,498 - INFO - db: test1, executing command: reindex index public."test_id1 's ";
2023-10-20 17:03:47,511 - INFO - db: test1, executing command: reindex index public.idx_projecttag;
2023-10-20 17:03:47,524 - INFO - db: test1, executing command: reindex index public." test_id "" 3 ";
2023-10-20 17:03:47,537 - INFO - db: test1, executing command: reindex index public."test_id2 \ $ \\";
2023-10-20 17:03:47,551 - INFO - db: test1, executing command: reindex index public.id1;
2023-10-20 17:03:47,560 - INFO - db: test3, total have 7 commands to execute
2023-10-20 17:03:47,563 - INFO - db: test3, executing command: reindex index test_idx_citext;
2023-10-20 17:03:47,578 - INFO - db: test3, executing command: reindex index test_citext_pkey;
2023-10-20 17:03:47,592 - INFO - db: test3, executing command: reindex index "test_id2 \ $ \\";
2023-10-20 17:03:47,605 - INFO - db: test3, executing command: reindex index " test_id "" 3 ";
2023-10-20 17:03:47,618 - INFO - db: test3, executing command: reindex index "test_id1 's ";
2023-10-20 17:03:47,631 - INFO - db: test3, executing command: reindex index id1;
2023-10-20 17:03:47,643 - INFO - db: test3, executing command: reindex index idx_projecttag;
2023-10-20 17:03:47,652 - INFO - db: test2, total have 7 commands to execute
2023-10-20 17:03:47,656 - INFO - db: test2, executing command: reindex index foo.id1;
2023-10-20 17:03:47,669 - INFO - db: test2, executing command: reindex index foo.test_idx_citext;
2023-10-20 17:03:47,683 - INFO - db: test2, executing command: reindex index foo." test_id "" 3 ";
2023-10-20 17:03:47,696 - INFO - db: test2, executing command: reindex index foo.test_citext_pkey;
2023-10-20 17:03:47,709 - INFO - db: test2, executing command: reindex index foo.idx_projecttag;
2023-10-20 17:03:47,722 - INFO - db: test2, executing command: reindex index foo."test_id2 \ $ \\";
2023-10-20 17:03:47,735 - INFO - db: test2, executing command: reindex index foo."test_id1 's ";
2023-10-20 17:03:47,743 - INFO - db: postgres, total have 3 commands to execute
2023-10-20 17:03:47,747 - INFO - db: postgres, executing command: reindex index pg_seclabel_object_index;
2023-10-20 17:03:47,760 - INFO - db: postgres, executing command: reindex index pg_shseclabel_object_index;
2023-10-20 17:03:47,772 - INFO - db: postgres, executing command: reindex index public.users_pkey;
2023-10-20 17:03:47,781 - INFO - All done
```

kainwen commented:
@bmdoil @kalensk Hi, any new comments? Thanks.

kalensk commented:
> Hi, any new comments? Thanks.

No new comments from me. Seems good overall to me.

kmacoskey commented:
Quick comment. Can we confirm that when you run through whole build process this file will actually be copied into the install directory? There might be a change needed in a Makefile somewhere to ensure that happens.

bmdoil reviewed (approved):
Overall LGTM, only remaining question is re. the build system https://github.com/greenplum-db/gpdb/pull/16312/files/ff563efbd65cb5318f916533213ca9407a897148#r1362827571

z-wenlin commented:
Thanks all for the review, the Makefile has been modified and it will install the script into the @bindir@.
